File: python/프로그래머스/피로도.py
# def solution(k, dungeons):
#     length = len(dungeons)
#     visited = [False for _ in range(length)]
    
#     def dfs(k, result):
#         max_result = result
        
#         for i in range(length):
#             if not visited[i] and k >= dungeons[i][0]:  # 최소 필요 피로도 확인
#                 visited[i] = True
#                 max_result = max(max_result, dfs(k - dungeons[i][1], result + 1))  # 탐색 결과 갱신
#                 visited[i] = False  # 백트래킹
        
#         return max_result

#     max_answer = 0
#     for i in range(length):
#         if k >= dungeons[i][0]:  # 시작 가능 여부 확인
#             visited[i] = True
#             max_answer = max(max_answer, dfs(k - dungeons[i][1], 1))  # 초기 탐색 시작
#             visited[i] = False

#     return max_answer

def solution(k, dungeons):
    answer = -1
    length = len(dungeons)
    visited = [False for _ in range(length)]
    for i in range(length):
        mini = dungeons[i][0]
        if mini <= k:
            visited[i] = True
            answer = max(answer,dfs(dungeons,length,k-dungeons[i][1],visited,1));
            visited[i] = False
    return answer

def dfs(dungeons,length,k,visited,result):
    max_result = result
    
    for i in range(length):
        if visited[i] == False:
            if k >= dungeons[i][0]:
                visited[i] = True
                max_result = max(max_result,dfs(dungeons,length,k-dungeons[i][1],visited,result + 1))
                visited[i] = False
    
    return max_result

File: java/피로도.java
import java.util.*;

class Solution {
    int maxK = -1; 
    int maxD;
    boolean[] visited;
    public int solution(int k, int[][] dungeons) {
        maxD = dungeons.length;
        visited = new boolean[maxD];
        
        for(int i=0; i<maxD;i++) {
            visited[i] = true;
            dfs(i, 0, k, dungeons);
            visited[i] = false;
            if(maxK == maxD) break;
        } 
        return maxK;
    }
    
    void dfs(int cnt, int count, int k, int[][] dungeons) {
        int minP = dungeons[cnt][0];
        if(minP > k) {
            maxK =Math.max(count, maxK);
            return;
        } 
        if(count+1 == maxD) {
            maxK = maxD;
            return;
        }
        k -= dungeons[cnt][1];
        for(int i=0;i<maxD;i++) {
            if(!visited[i]) {
                visited[i] = true;
                dfs(i, count+1, k, dungeons);
                visited[i] = false;
            }
        }
    }
}
